Implement RFC 021 Projections api

George-Payne opened the pull request:
**This PR implements only the parts of RFC021 currently supported via GRPC.**

- Create:
	- Rename `createContinuousProjection` to `createProjection`
	- Remove `createOneTimeProjection`
	- Remove `createTransientProjection`
- List:
	- Rename `listContinuousProjection` to `listProjection`
	- Remove `listOneTimeProjection`
	- Remove `listTransientProjection`
- Details:
	- remove `mode` from `ProjectionDetails`
	- delete `CONTINUOUS` `ONE_TIME` & `TRANSIENT` constants
- Update:
	- Rename `trackEmittedStreams` to `emitEnabled` (Was incorrectly named, `trackEmittedStreams` to be added in future version.)
- Reset:
	- remove `writeCheckpoint` option (had no effect)
- Disable:
	- In `disableProjection` remove `writeCheckpoint` option. (Always true)
	- Add `abortProjection` method. (`writeCheckpoint` always false)
- Tests:
	- Update tests to match updated behaviour
	- Gate fuzzy disable test matching behind server version, for stricter tests on known good server implementations
- Samples:
	- Update samples to match new API


RFC: https://github.com/EventStore/architecture-and-planning/blob/main/rfcs/021%20-%20gRPC%20Projections.md

oskardudycz commented on src/projections/updateProjection.ts:
```diff
@@ -8,10 +8,10 @@ import { debug, convertToCommandError } from "../utils";
 export interface UpdateProjectionOptions extends BaseOptions {
   /**
-   * Enables tracking emitted streams.
+   * Enables emitting events from the projection.
```
**CONCERN:** I'm not sure if that's in the scope of this PR, but `Enabled` parameter is missing defining if projection should be enabled by default.

oskardudycz replied:
Hm, after rereading the RFC, it might be fine. RFC states at first that there should be such a parameter:
![obraz](https://user-images.githubusercontent.com/5562528/142173028-46f3e40f-4120-414a-8139-695057c384ff.png)
But then it states that it should be removed:
> Remove the Enabled parameter from CreateProjection. This will always be false, and the projection must be enabled as a separate step with the Enable operation on the proto.
@hayley-jean thoughts?

oskardudycz commented on src/projections/updateProjection.ts:
```diff
@@ -8,10 +8,10 @@ import { debug, convertToCommandError } from "../utils";
    * @default undefined
    */
-  trackEmittedStreams?: boolean | undefined;
+  emitEnabled?: boolean | undefined;
```
**QUESTION:** Why do we have `| undefined` if it's optional? Shouldn't we just have?
```suggestion
  emitEnabled?: boolean;
```

George-Payne replied:
Yeah, probably, it means the same thing. 

I think my thinking back when I originally wrote that was that `undefined` is an acceptable value, and has meaning, and I wanted to imply that via the type.  e.g. "emit enabled can have the value of boolean or undefined. Its also optional to pass this."

George-Payne replied:
Removed, and updated the doc comment to explicitly say what undefined means.

oskardudycz commented on src/projections/updateProjection.ts:
```diff
@@ -8,10 +8,10 @@ import { debug, convertToCommandError } from "../utils";
 
 export interface UpdateProjectionOptions extends BaseOptions {
```
**COMMENT:** To have RFC fully implemented, `TrackEmittedStreams` has to be added.

George-Payne replied:
Indeed, this PR only covers the parts of the RFC currently supported by the server over GRPC (`TrackEmittedStreams` is missing in the protos). I should probably update the wording of the PR to mention that (the decision to "polyfill" the gaps was made after this PR).

oskardudycz commented:
Based on the updates in https://github.com/EventStore/architecture-and-planning/pull/46/files. Some implementation will need to be aligned, like adding missing methods: `GetConfig`, `GetQuery`, `GetStatus`, etc. Still, this can be handled in a separate PR.

I think that setting default values to Delete projection parameters to `false` could be included here (see: https://github.com/EventStore/architecture-and-planning/pull/46/files#diff-1af18be3bf99a6d073701661e1e26f3722ac2a936aa260544f5b4b2cbffbd923R43).

George-Payne commented:
> Adding missing methods: `GetConfig`, `GetQuery`, `GetStatus`, etc. Still, this can be handled in a separate PR.

I've updated the PR body to mention that this PR only covers the parts currently supported over GRPC.

 > I think that setting default values to Delete projection parameters to `false` could be included here
 
 Good spot, updated.

oskardudycz commented:
@George-Payne, good to go from my end 👍
